nav2_amcl/src/amcl_node.cpp

```diff
@@ -531,13 +311,8 @@ AmclNode::globalLocalizationCallback(
   const std::shared_ptr<std_srvs::srv::Empty::Request>/*req*/,
   std::shared_ptr<std_srvs::srv::Empty::Response>/*res*/)
 {
-  if (map_ == NULL) {
```

Comment:
So with this, whenever we want to update the map we have to bring the node to un-configured state and then we need to re-configure everything. Which means we will have to re-initialize particle filter, pose, and everything else.  This way we will lose track of previous poses and other information unless during this transition we save/reload necessary system states.

Reply:
It comes down to what is meant by each of the Lifecycle states. When a node is first created it is **Unconfigured**. Then one configures a node and it becomes **Inactive** (configured, but not yet active). Then, the node is activated and it enters the **Active** state. 

From the ROS2 Design:

> Transition State: Configuring
> In this transition state the node’s onConfigure callback will be called to allow the node to load its configuration and conduct any required setup.
> 
> The configuration of a node will typically involve those tasks that must be performed once during the node’s life time, such as obtaining permanent memory buffers and setting up topic publications/subscriptions that do not change.
> 
> The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.

> Transition State: Activating
> In this transition state the callback onActivate will be executed. This method is expected to do any final preparations to start executing. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.

For the initial implementation of lifecycle nodes in the Navigation 2 stack, I've interpreted it this way:

* Configure: read its parameters, allocate pubs/subs, allocate buffers, etc. 
* Activate: activate its publishers, perhaps launch threads; do whatever remains to get the node up and running. 

This simple approach (reading all of the parameters during on_configure) does mean that to reconfigure a node one must currently deactivate and cleanup and then configure and activate. However, going forward, I plan to look at every parameter and decide whether it should be allowed to be read during on_activate, or even whether it should be allowed to be changed on the fly, during the active state (this should be rare as it could impact real-time performance). 

Once we start associating parameters with states in which they can be changed, there is then a new issue about having to get the node into the right state for the parameter change. I think that for every parameter, we'll need to associate which state the node needs to be in to update the parameter. This should be done in such a way that we can interface to the Nav2 system with a generic parameter updating UI.